lib/fxpay/errors.js

```diff
@@ -0,0 +1,75 @@
+(function() {
+  'use strict';
+
+  var exports = fxpay.errors = {};
+
+  // All error classes will implicitly inherit from this.
+  exports.FxPayError = createError('FxPayError');
+
+
+  exports.AddReceiptError = createError('AddReceiptError');
+  exports.AppOriginMissing = createError('AppOriginMissing');
+  exports.AppReceiptMissing = createError('AppReceiptMissing');
+  exports.FailedWindowMessage = createError('FailedWindowMessage');
+  exports.IncorrectUsage = createError('IncorrectUsage');
+  exports.InvalidApp = createError('InvalidApp');
+  exports.InvalidJWTData = createError('InvalidJWTData');
+  exports.InvalidPayProviderURL = createError('InvalidPayProviderURL');
+  exports.InvalidReceipt = createError('InvalidReceipt');
+  exports.InvalidTransactionState = createError('InvalidTransactionState');
+  exports.MissingPaymentWindow = createError('MissingPaymentWindow');
+  exports.NotImplementedError = createError('NotImplementedError');
+  exports.PayPlatformError = createError('PayPlatformError');
+  exports.PayPlatformUnavailable = createError('PayPlatformUnavailable');
+  exports.PayWindowClosedByUser = createError('PayWindowClosedByUser',
+                                              {code: 'DIALOG_CLOSED_BY_USER'});
+  exports.TestReceiptNotAllowed = createError('TestReceiptNotAllowed');
+  exports.TransactionTimeout = createError('TransactionTimeout');
+  exports.UnexpectedJWTType = createError('UnexpectedJWTType');
+  exports.UndefinedAppOrigin = createError('UndefinedAppOrigin');
+  exports.UnknownMessageOrigin = createError('UnknownMessageOrigin');
+  exports.UnknownMessageStatus = createError('UnknownMessageStatus');
+  exports.WrongJWTSegmentCount = createError('WrongJWTSegmentCount');
+
+
+  exports.APIError = createError('APIError');
+
+  var apiClassOpt = {inherits: exports.APIError};
+  exports.APIRequestAborted = createError('APIRequestAborted', apiClassOpt);
+  exports.APIRequestError = createError('APIRequestError', apiClassOpt);
+  exports.APIRequestTimeout = createError('APIRequestTimeout', apiClassOpt);
+  exports.BadAPIResponse = createError('BadAPIResponse', apiClassOpt);
+  exports.BadJSONResponse = createError('BadJSONResponse', apiClassOpt);
+
+
+  function createError(name, classOpt) {
+    classOpt = classOpt || {};
+    var errorParent = classOpt.inherits || exports.FxPayError || Error;
+
+    function _FxPayError(message, opt) {
+      opt = opt || {};
+      /* jshint -W040 */
+      if (typeof this === 'undefined') {
```

Comment:
Would be nice to use this pattern here: 
```javascript
function MyConstructor(){
  if (!(this instanceof MyConstructor)){
    return new MyConstructor();
  }
  this.bar = 'MyProperty';
}
```


See it in action here: http://jsbin.com/levoxidite/1/edit?js,console . That way it'll work with and without the `new` and you can drop the (imho ugly) new from all the usage.